src/lib/comms/sol-bluetooth-impl-linux.c

```diff
@@ -0,0 +1,683 @@
+/*
+ * This file is part of the Soletta Project
+ *
+ * Copyright (C) 2016 Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <systemd/sd-bus.h>
+
+#include <sol-bus.h>
+#include <sol-log.h>
+#include <sol-network.h>
+#include <sol-bluetooth.h>
+#include <sol-util-internal.h>
+
+struct sol_bt_conn {
+    int ref;
+};
+
+struct device_info {
+    char *path;
+    uint64_t mask;
+    struct sol_bt_device_info info;
+};
+
+struct sol_bt_scan_pending {
+    sd_bus_slot *slot;
+    void (*callback)(void *user_data, const struct sol_bt_device_info *device);
+    void *user_data;
+};
+
+static struct context {
+    sd_bus *system_bus;
+    struct sol_bus_client *bluez;
+    char *adapter_path;
+    struct sol_ptr_vector devices;
+    void (*enabled_callback)(void *user_data, bool powered);
+    struct sol_bt_scan_pending pending_scan;
+    void *enabled_user_data;
+    bool adapter_powered;
+} context;
+
+enum {
+    ADAPTER_PROPERTY_POWERED = 0,
+};
+
+static void
+strv_free(char **strv)
+{
+    char **l;
+    SOL_NULL_CHECK(strv);
+
+    for (l = strv; *l; l++)
+        free(*l);
+
+    free(strv);
+}
+
+static bool
+adapter_property_powered_set(void *data, const char *path, sd_bus_message *m)
+{
+    struct context *ctx = data;
+    bool powered;
+    int r;
+
+    r = sd_bus_message_read_basic(m, SD_BUS_TYPE_BOOLEAN, &powered);
+    SOL_INT_CHECK_GOTO(r, < 0, skip);
+
+    return ctx->adapter_powered != powered;
+
+skip:
+    r = sd_bus_message_skip(m, "b");
+    SOL_INT_CHECK(r, < 0, false);
+
+    return false;
+}
+
+static const struct sol_bus_properties adapter_properties[] = {
+    [ADAPTER_PROPERTY_POWERED] = {
+        .member = "Powered",
+        .set = adapter_property_powered_set,
+    },
+    { NULL, NULL }
+};
+
+static void
+adapter_property_changed(void *data, const char *path, uint64_t mask)
+{
+    struct context *ctx = data;
+
+    if (mask & (1 << ADAPTER_PROPERTY_POWERED)) {
+        ctx->adapter_powered = !ctx->adapter_powered;
```

Comment:
Yeah, this ended being more complicated than it needs to be.

This is how it works: `adapter_property_powered_set()` only returns `true` when the adapter powered state is different from the stored one, `_set()` returning `true` causes this function to be called with the appropriate mask. So I know that the powered is state is different than the one before.